reactor-core/src/main/java/reactor/core/publisher/InternalManySink.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2011-Present VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import reactor.core.Exceptions;
+
+interface InternalManySink<T> extends Sinks.Many<T>, ContextHolder {
+
+	@Override
+	default void emitNext(T value, Sinks.EmitStrategy strategy) {
+		for (;;) {
+			Sinks.Emission emission = tryEmitNext(value);
+			if (emission.hasSucceeded()) {
+				return;
+			}
+
+			boolean shouldRetry = strategy.onEmissionFailure(emission);
+			if (shouldRetry) {
+				continue;
+			}
+
+			switch (emission) {
+				case FAIL_ZERO_SUBSCRIBER:
+					//we want to "discard" without rendering the sink terminated.
+					// effectively NO-OP cause there's no subscriber, so no context :(
+					return;
+				case FAIL_OVERFLOW:
+					Operators.onDiscard(value, currentContext());
+					//the emitError will onErrorDropped if already terminated
+					emitError(Exceptions.failWithOverflow("Backpressure overflow during Sinks.Many#emitNext"));
+					return;
+				case FAIL_CANCELLED:
+					Operators.onDiscard(value, currentContext());
+					return;
+				case FAIL_TERMINATED:
+					Operators.onNextDropped(value, currentContext());
+					return;
+				case FAIL_NON_SERIALIZED:
+					throw new IllegalStateException("Non-serialized access");
```

Comment:
I suspect most users will be really surprised by the `throw`, considering that's something we go out of our way to avoid in reactor... so that possibility should at least be documented in the javadoc... since we have a `EmissionException`, I'd use that (so that it can be caught specifically).